From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aviana Cruz <gwencroft@proton.me>
Date: Fri, 23 Dec 2022 15:05:15 +0800
Subject: [PATCH] Backport ArrowDMG fix

Upstream commit: PaperMC/Paper@bb397ba74c034ffe5ae164d774a043cb7bca0472

diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index d558f72771c9b7c64d4f2c545e0b5529bd72f718..ed416aab3398e57553a4e611f7863d048348e31e 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1719,13 +1719,26 @@ public abstract class EntityLiving extends Entity {
         return 0.42F;
     }
 
+    protected long lastJumpTime = 0L; // Dionysus - Backport ArrowDMG fix
     protected void cu() {
+        // Dionysus start - Backport ArrowDMG fix
+        long time = System.nanoTime();
+        boolean canCrit = true;
+        if (this instanceof EntityPlayer) {
+            canCrit = false;
+            if (time - this.lastJumpTime > (long)(0.250e9)) {
+                this.lastJumpTime = time;
+                canCrit = true;
+            }
+        }
+        // Dionysus end - Backport ArrowDMG fix
+
         this.motY = (double) this.ct();
         if (this.hasEffect(MobEffects.JUMP)) {
             this.motY += (double) ((float) (this.getEffect(MobEffects.JUMP).getAmplifier() + 1) * 0.1F);
         }
 
-        if (this.isSprinting()) {
+        if (canCrit && this.isSprinting()) { // Dionysus - Backport ArrowDMG fix
             float f = this.yaw * 0.017453292F;
 
             this.motX -= (double) (MathHelper.sin(f) * 0.2F);
